sdk/core/azure_core_macros/src/tracing_new.rs

```diff
@@ -0,0 +1,312 @@
+use quote::{quote, ToTokens};
+use syn::{parse::Parse, spanned::Spanned, ExprStruct, ItemFn, Result};
+
+const INVALID_SERVICE_CLIENT_NEW_MESSAGE: &str =
+    "new attribute must be applied to a public function with a name starting with `new`";
+
+fn parse_struct_expr(
+    client_namespace: &str,
+    struct_body: &ExprStruct,
+    default: TokenStream,
+    is_ok: bool,
+) -> TokenStream {
+    if struct_body.path.is_ident("Self") {
+        let fields = struct_body.fields.iter();
+        let before_self = quote! {
+            let tracer =
+            if let Some(tracer_options) = &options.client_options.request_instrumentation {
+                tracer_options
+                    .tracer_provider
+                    .as_ref()
+                    .map(|tracer_provider| {
+                        tracer_provider.get_tracer(
+                            Some(#client_namespace),
+                            option_env!("CARGO_PKG_NAME").unwrap_or("UNKNOWN"),
+                            option_env!("CARGO_PKG_VERSION").unwrap_or("UNKNOWN"),
+                        )
+                    })
+            } else {
+                None
+            };
+        };
+        if is_ok {
+            quote! {
+                #before_self
+                Ok(Self {
+                    tracer,
+                    #(#fields),*,
+                })
+            }
+        } else {
+            quote! {
+                #before_self
+                Self {
+                    tracer,
+                    #(#fields),*,
+                }
+            }
+        }
+    } else {
+        println!("ident is not Self, emitting expression: {default:?}");
+        default
+    }
+}
+
+struct NamespaceAttribute {
+    client_namespace: String,
+}
+
+impl Parse for NamespaceAttribute {
+    fn parse(input: syn::parse::ParseStream) -> Result<Self> {
+        let client_namespace = input.parse::<syn::LitStr>()?.value();
+        Ok(NamespaceAttribute { client_namespace })
+    }
+}
+
+/// Parse the token stream for an Azure Service client "new" declaration.
+///
+/// An Azure Service client "new" declaration is a public function whose name starts with
+/// `new` and returns either a new client instance or an error.
+///
+/// This macro will ensure that the fn is public and returns one of the following:
+/// 1) `Self`
+/// 1) `Arc<Self>`
+/// 1) `Result<Self, E>`
+/// 1) `Result<Arc<Self>, E>`
+///
+pub fn parse_new(attr: TokenStream, item: TokenStream) -> Result<TokenStream> {
+    if !is_new_declaration(&item) {
+        return Err(syn::Error::new(
+            item.span(),
+            INVALID_SERVICE_CLIENT_NEW_MESSAGE,
+        ));
+    }
+    let namespace_attrs: NamespaceAttribute = syn::parse2(attr)?;
+
+    let client_fn: ItemFn = syn::parse2(item.clone())?;
+
+    let vis = &client_fn.vis;
+    let ident = &client_fn.sig.ident;
+    let inputs = client_fn.sig.inputs.iter();
+    let body = client_fn.block.stmts.iter().map(|stmt| {
```

Comment:
Should we track if we found a suitable statement to rewrite and emit an error if we found no such statement? If we don't find a statement to rewrite, the attribute didn't do anything, which feels like an error. The user should just remove the attribute.